File: src/pipeline.cpp
#include "doe/pipeline.hpp"

#include "doe/energy.hpp"
#include "doe/image.hpp"
#include "doe/init.hpp"
#include "doe/targets.hpp"

#include <chrono>
#include <cmath>
#include <format>
#include <stdexcept>

namespace doe {

Grid grid_for(const DesignConfig& cfg) {
    return Grid::padded_for(cfg.active, cfg.pitch, cfg.wavelength, cfg.distance, cfg.pad);
}

SamplingReport sampling_for(const DesignConfig& cfg) {
    return sampling_report(grid_for(cfg), cfg.distance, cfg.active * cfg.pitch);
}

DesignInputs prepare(const Array2<double>& target, const DesignConfig& cfg) {
    if (cfg.active < 2) throw std::invalid_argument("prepare: active aperture must be >= 2 px");
    DesignInputs in;
    in.grid = grid_for(cfg);
    const std::size_t n = static_cast<std::size_t>(in.grid.n), a = static_cast<std::size_t>(cfg.active);
    // Keep the target's aspect ratio: the longer side fills the aperture, the
    // window is the resampled rectangle. The DOE stays square.
    const double scale = double(a) / double(std::max(target.rows, target.cols));
    const std::size_t ar = target.rows >= target.cols ? a : std::max<std::size_t>(1, static_cast<std::size_t>(std::lround(target.rows * scale)));
    const std::size_t ac = target.cols >= target.rows ? a : std::max<std::size_t>(1, static_cast<std::size_t>(std::lround(target.cols * scale)));
    Array2<double> t = resample(target, ar, ac);
    if (cfg.soft_edge_sigma > 0.0) t = targets::soft_edges(t, cfg.soft_edge_sigma);
    double tmax = 0.0;
    for (double v : t.data) tmax = std::max(tmax, v);
    if (tmax > 0.0)
        for (double& v : t.data) v = std::max(v, 0.0) / tmax;
    in.i_target = embed(t, n);
    in.b = in.i_target;
    for (double& v : in.b.data) v = std::sqrt(v);

    // Metrics always on the image rectangle; the energy window is the whole aperture in
    // letterbox mode (the strips are dark targets) or the rectangle otherwise.
    in.metrics_mask = embed(Array2<double>(ar, ac, 1.0), n);
    in.mask = cfg.letterbox ? embed(Array2<double>(a, a, 1.0), n) : in.metrics_mask;
    Array2<double> il(a, a, 0.0);
    const double c = (double(a) - 1.0) / 2.0, half = double(a) / 2.0;
    for (std::size_t i = 0; i < a; ++i)
        for (std::size_t j = 0; j < a; ++j) {
            const double x = double(i) - c, y = double(j) - c, r = std::hypot(x, y);
            switch (cfg.illum) {
                case IllumShape::square: il(i, j) = 1.0; break;
                case IllumShape::disk: il(i, j) = r <= half ? 1.0 : 0.0; break;
                case IllumShape::gaussian: {
                    const double w = cfg.gaussian_waist * half;
                    il(i, j) = r <= half ? std::exp(-(r * r) / (w * w)) : 0.0;
                    break;
                }
            }
        }
    in.illum = embed(il, n);
    return in;
}

namespace {

template <class T>
Array2<T> cast(const Array2<double>& a) {
    Array2<T> out(a.rows, a.cols);
    for (std::size_t k = 0; k < a.data.size(); ++k) out.data[k] = T(a.data[k]);
    return out;
}
Array2<double> to_double(const Array2<float>& a) {
    Array2<double> out(a.rows, a.cols);
    for (std::size_t k = 0; k < a.data.size(); ++k) out.data[k] = a.data[k];
    return out;
}
Array2<double> to_double(const Array2<double>& a) { return a; }
Field<double> to_double(const Field<float>& a) {
    Field<double> out(a.rows, a.cols);
    for (std::size_t k = 0; k < a.data.size(); ++k) out.data[k] = std::complex<double>(a.data[k]);
    return out;
}
Field<double> to_double(const Field<double>& a) { return a; }

template <class T>
void run_solvers(const DesignConfig& cfg, DesignResult& r, const DesignProgress& progress) {
    const DesignInputs& in = r.inputs;
    AngularSpectrum<T> prop(in.grid, cfg.distance, cfg.band_limit);
    const Array2<T> illum = cast<T>(in.illum), b = cast<T>(in.b), mask = cast<T>(in.mask), mmask = cast<T>(in.metrics_mask);
    const Array2<T> phi0 = cast<T>(r.phi0);
    const EnergyWeights weights{1.0, cfg.mu};

    r.initial_metrics = metrics(energy_and_grad<T>(phi0, illum, prop, b, mask, weights).field, b, mmask);

    auto timed = [&](const std::string& name, auto&& fn) {
        const auto t0 = std::chrono::steady_clock::now();
        Result<T> res = fn();
        SolverRun run;
        run.name = name;
        run.seconds = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
        run.phi = to_double(res.phi);
        run.field = to_double(res.field);
        run.history = std::move(res.history);
        run.shape_history = std::move(res.shape_history);
        run.efficiency_history = std::move(res.efficiency_history);
        run.metrics = metrics(res.field, b, mmask);
        r.runs.push_back(std::move(run));
    };
    auto cb = [&](const std::string& name) -> ProgressFn {
        if (!progress) return {};
        return [&, name](int it, double e) { return progress(name, it, e); };
    };

    if (cfg.run_gs) {
        GsParams gp = cfg.gs;
        gp.weights = weights;  // the logged history uses the same mu as the Adam runs
        timed("gs", [&] { return gerchberg_saxton<T>(phi0, illum, prop, b, mask, gp); });
    }

    OptimizeParams op;
    op.iters = cfg.iters;
    op.lr = cfg.lr;
    op.weights = weights;
    timed("adam", [&] { return optimize<T>(phi0, illum, prop, b, mask, op, cb("adam")); });

    if (cfg.levels > 0) {
        OptimizeParams oq = op;
        oq.quant.levels = cfg.levels;
        oq.quant.method = cfg.quant_method;
        oq.quant.start = cfg.quant_start;
        oq.quant.ramp = cfg.quant_ramp;
        oq.quant.choi_w = cfg.choi.choi_w;
        oq.quant.choi_gain0 = cfg.choi.choi_gain0;
        oq.quant.choi_gain1 = cfg.choi.choi_gain1;
        oq.quant.choi_tau0 = cfg.choi.choi_tau0;
        oq.quant.choi_c = cfg.choi.choi_c;
        oq.quant.choi_seed = cfg.seed;
        const std::string name = std::format("adam_q{}", cfg.levels);
        timed(name, [&] { return optimize<T>(phi0, illum, prop, b, mask, oq, cb(name)); });
    }
}

}  // namespace

DesignResult design(const Array2<double>& target, const DesignConfig& cfg, const DesignProgress& progress) {
    DesignResult r;
    r.config = cfg;
    r.inputs = prepare(target, cfg);
    r.sampling = sampling_for(cfg);
    if (cfg.init == InitMethod::tie) {
        Array2<double> i_source = r.inputs.illum;
        for (double& v : i_source.data) v *= v;
        r.phi0 = init_tie<double>(r.inputs.grid, cfg.distance, r.inputs.i_target, i_source);
    } else if (cfg.init == InitMethod::backprop) {
        AngularSpectrum<double> prop(r.inputs.grid, cfg.distance, cfg.band_limit);
        r.phi0 = init_backprop<double>(r.inputs.grid, prop, r.inputs.b);
    } else {
        r.phi0 = init_random<double>(r.inputs.grid, cfg.seed);
    }
    if (cfg.single_precision) run_solvers<float>(cfg, r, progress);
    else run_solvers<double>(cfg, r, progress);
    return r;
}

Json DesignResult::report() const {
    Json j;
    Json c;
    c.set("wavelength_m", config.wavelength);
    c.set("pitch_m", config.pitch);
    c.set("distance_m", config.distance);
    c.set("active_px", config.active);
    c.set("padded_px", inputs.grid.n);
    c.set("band_limit", config.band_limit);
    c.set("letterbox", config.letterbox);
    c.set("single_precision", config.single_precision);
    c.set("illumination", config.illum == IllumShape::square ? "square" : config.illum == IllumShape::disk ? "disk" : "gaussian");
    c.set("init", config.init == InitMethod::tie ? "tie" : config.init == InitMethod::random ? "random" : "backprop");
    c.set("seed", static_cast<int>(config.seed));
    c.set("iters", config.iters);
    c.set("lr", config.lr);
    c.set("mu", config.mu);
    c.set("levels", config.levels);
    c.set("quant_method", config.quant_method == QuantMethod::wyrowski ? "wyrowski" : "choi");
    c.set("quant_start", config.quant_start);
    c.set("quant_ramp", config.quant_ramp == QuantRamp::linear ? "linear" : "table");
    c.set("gs_iters", config.gs.iters);
    c.set("gs_phase_only_iters", config.gs.phase_only_iters);
    c.set("soft_edge_sigma_px", config.soft_edge_sigma);
    j.set("config", c);

    Json s;
    s.set("max_angle_deg", sampling.max_angle_rad * 180.0 / 3.14159265358979323846);
    s.set("spot_size_m", sampling.spot_size_m);
    s.set("spot_size_px", sampling.spot_size_px);
    s.set("signal_spread_m", sampling.signal_spread_m);
    s.set("window_extent_m", sampling.window_extent_m);
    s.set("fresnel_number", sampling.fresnel_number);
    s.set("wrap_misses_picture", sampling.wrap_misses_picture);
    s.set("spot_resolved", sampling.spot_resolved);
    std::vector<std::string> warnings;
    if (!sampling.wrap_misses_picture) warnings.push_back("light wrapping around the periodic window reaches the picture: enlarge the window, reduce the distance or the pitch");
    if (!sampling.spot_resolved) warnings.push_back("diffraction-limited spot is larger than one target pixel: finer detail is unreachable");
    s.set("warnings", warnings);
    j.set("sampling", s);

    const double pitch_m = config.pitch;
    auto metrics_json = [pitch_m](const Metrics& m) {
        Json mj;
        mj.set("efficiency", m.efficiency);
        mj.set("amplitude_rmse", m.amplitude_rmse);
        mj.set("speckle_contrast", m.speckle_contrast);
        mj.set("ncc", m.ncc);
        mj.set("psnr_db", m.psnr_db);
        mj.set("vortex_count", static_cast<int>(m.vortex_count));
        mj.set("vortex_count_bright", static_cast<int>(m.vortex_count_bright));
        mj.set("vortex_density_bright_per_mm2", m.vortex_density_bright / (pitch_m * pitch_m) * 1e-6);
        return mj;
    };
    j.set("initial", metrics_json(initial_metrics));
    Json rj;
    for (const auto& run : runs) {
        Json one;
        one.set("metrics", metrics_json(run.metrics));
        one.set("iterations", static_cast<int>(run.history.size()));
        one.set("seconds", run.seconds);
        one.set("final_energy", run.history.empty() ? 0.0 : run.history.back());
        rj.set(run.name, one);
    }
    j.set("runs", rj);
    return j;
}

std::string DesignResult::markdown_table() const {
    const double per_mm2 = 1e-6 / (config.pitch * config.pitch);
    std::string s = "| run | efficiency | amplitude RMSE | speckle contrast | NCC | PSNR (dB) | vortices in bright /mm² | iterations | time (s) |\n|---|---|---|---|---|---|---|---|---|\n";
    s += std::format("| initial ({}) | {:.4f} | {:.4f} | {:.3f} | {:.4f} | {:.2f} | {:.0f} | - | - |\n",
                     config.init == InitMethod::tie ? "TIE" : config.init == InitMethod::random ? "random" : "backprop", initial_metrics.efficiency, initial_metrics.amplitude_rmse,
                     initial_metrics.speckle_contrast, initial_metrics.ncc, initial_metrics.psnr_db, initial_metrics.vortex_density_bright * per_mm2);
    for (const auto& r : runs)
        s += std::format("| {} | {:.4f} | {:.4f} | {:.3f} | {:.4f} | {:.2f} | {:.0f} | {} | {:.1f} |\n", r.name, r.metrics.efficiency,
                         r.metrics.amplitude_rmse, r.metrics.speckle_contrast, r.metrics.ncc, r.metrics.psnr_db,
                         r.metrics.vortex_density_bright * per_mm2, r.history.size(), r.seconds);
    return s;
}

}  // namespace doe

File: viewer-web/core/pipeline.js
// Design pipeline, ported from src/pipeline.cpp and src/render.cpp: prepare (bilinear resampling
// with the aspect ratio kept; zero embedding; square / disk / Gaussian
// illumination), design (start, optional GS baseline, Adam, optional quantized Adam, metrics,
// sampling report), sweepVolume (planes between the DOE and the target, cuts, vortex statistics)
// and attachSource (format v2 source block). Targets are { rows, cols, data } with i along x.
(function () {
  // Module factory. As a classic script it also registers its own source so that the page can
  // build a Web Worker from it (a worker script cannot be loaded from a file:// URL).
  function factory(root) {
  const isNode = typeof module !== 'undefined' && module.exports;
  const C = isNode ? Object.assign({}, require('./propagate.js'), require('./energy.js'), require('./init.js'), require('./metrics.js'), require('./solvers.js'), require('./util.js')) : root.DOE_CORE;
  const { Grid, nextSmooth, nextPow2, AngularSpectrum, samplingReport, energyAndGrad, initTie, initBackprop, initRandom, metrics, optimize, gerchbergSaxton, vortexChargeMap, vortexCount, wrapToPi } = C;

  function resample(src, rows, cols) {
    if (!(src.rows > 0 && src.cols > 0 && rows > 0 && cols > 0) || src.data.length !== src.rows * src.cols) throw new Error('resample: empty array or size mismatch');
    const out = new Float64Array(rows * cols);
    const sx = rows > 1 ? (src.rows - 1) / (rows - 1) : 0, sy = cols > 1 ? (src.cols - 1) / (cols - 1) : 0;
    for (let i = 0; i < rows; ++i) {
      const u = i * sx, i0 = Math.min(Math.floor(u), src.rows - 1), i1 = Math.min(i0 + 1, src.rows - 1), fu = u - i0;
      for (let j = 0; j < cols; ++j) {
        const v = j * sy, j0 = Math.min(Math.floor(v), src.cols - 1), j1 = Math.min(j0 + 1, src.cols - 1), fv = v - j0;
        out[i * cols + j] = (1 - fu) * ((1 - fv) * src.data[i0 * src.cols + j0] + fv * src.data[i0 * src.cols + j1]) + fu * ((1 - fv) * src.data[i1 * src.cols + j0] + fv * src.data[i1 * src.cols + j1]);
      }
    }
    return out;
  }
  function embed(data, rows, cols, n) {
    const out = new Float64Array(n * n), i0 = (n - rows) >> 1, j0 = (n - cols) >> 1;
    for (let i = 0; i < rows; ++i) for (let j = 0; j < cols; ++j) out[(i0 + i) * n + j0 + j] = data[i * cols + j];
    return out;
  }
  function cropCenter(data, n, view) {
    const out = new Float64Array(view * view), o = (n - view) >> 1;
    for (let i = 0; i < view; ++i) for (let j = 0; j < view; ++j) out[i * view + j] = data[(o + i) * n + o + j];
    return out;
  }

  const defaults = { active: 512, pitch: 8e-6, wavelength: 532e-9, distance: 0.05, illum: 'square', gaussianWaist: 0.5, init: 'tie', seed: 0,
    iters: 400, lr: 0.05, mu: 0.3, levels: 0, quantStart: 0.4, quantRamp: 'linear', runGs: true, gs: { iters: 40, phaseOnlyIters: 20 }, bandLimit: true, softEdge: 0,
    letterbox: true };   // non-square targets: square window with dark strips; false = rectangular window

  function prepare(target, cfgIn) {
    const cfg = Object.assign({}, defaults, cfgIn);
    if (cfg.active < 2) throw new Error('prepare: active aperture must be >= 2 px');
    if (!target || !(target.rows > 0 && target.cols > 0)) throw new Error('prepare: empty target image');
    // picture-clear rule; the browser page asks for power-of-two windows (pow2Grid) for the fast radix-2 FFT
    const grid = Grid.paddedFor(cfg.active, cfg.pitch, cfg.wavelength, cfg.distance, 1.0, cfg.pow2Grid ? nextPow2 : nextSmooth), n = grid.n, a = cfg.active;
    const scale = a / Math.max(target.rows, target.cols);
    const ar = target.rows >= target.cols ? a : Math.max(1, Math.round(target.rows * scale));
    const ac = target.cols >= target.rows ? a : Math.max(1, Math.round(target.cols * scale));
    const t = resample(target, ar, ac);
    let tmax = 0; for (const v of t) tmax = Math.max(tmax, v);
    if (tmax > 0) for (let k = 0; k < t.length; ++k) t[k] = Math.max(t[k], 0) / tmax;
    const iTarget = embed(t, ar, ac, n), b = iTarget.map(Math.sqrt);
    // metrics always on the image rectangle; the energy window is the whole aperture in
    // letterbox mode (the strips are dark targets) or the rectangle otherwise
    const metricsMask = embed(new Float64Array(ar * ac).fill(1), ar, ac, n);
    const mask = cfg.letterbox ? embed(new Float64Array(a * a).fill(1), a, a, n) : metricsMask;
    const il = new Float64Array(a * a), c = (a - 1) / 2, half = a / 2;
    for (let i = 0; i < a; ++i)
      for (let j = 0; j < a; ++j) {
        const r = Math.hypot(i - c, j - c);
        if (cfg.illum === 'square') il[i * a + j] = 1;
        else if (cfg.illum === 'disk') il[i * a + j] = r <= half ? 1 : 0;
        else { const w = cfg.gaussianWaist * half; il[i * a + j] = r <= half ? Math.exp(-(r * r) / (w * w)) : 0; }
      }
    return { grid, illum: embed(il, a, a, n), b, mask, metricsMask, iTarget, windowRows: ar, windowCols: ac, config: cfg };
  }

  // progress(name, it, energy) -> false cancels. Returns { config, inputs, sampling, phi0, initialMetrics, runs, canceled }.
  function design(target, cfgIn, progress) {
    const inputs = prepare(target, cfgIn), cfg = inputs.config, grid = inputs.grid, N = grid.n * grid.n;
    const prop = new AngularSpectrum(grid, cfg.distance, cfg.bandLimit);
    const r = { config: cfg, inputs, sampling: samplingReport(grid, cfg.distance, cfg.active * cfg.pitch), runs: [], canceled: false };
    if (cfg.init === 'tie') r.phi0 = initTie(grid, cfg.distance, inputs.iTarget, inputs.illum.map((x) => x * x));
    else if (cfg.init === 'backprop') r.phi0 = initBackprop(grid, prop, inputs.b);
    else r.phi0 = initRandom(grid, cfg.seed);
    const weights = { shape: 1, efficiency: cfg.mu };
    r.initialMetrics = metrics(energyAndGrad(r.phi0, inputs.illum, prop, inputs.b, inputs.mask, weights).field, inputs.b, inputs.metricsMask);
    const timed = (name, fn) => {
      const t0 = Date.now();
      const res = fn();
      r.runs.push({ name, phi: res.phi, field: res.field, history: res.history, shapeHistory: res.shapeHistory, efficiencyHistory: res.efficiencyHistory,
        metrics: metrics(res.field, inputs.b, inputs.metricsMask), seconds: (Date.now() - t0) / 1000 });
      return res;
    };
    const cb = (name) => (progress ? (it, e) => { const go = progress(name, it, e); if (go === false) r.canceled = true; return go; } : undefined);
    if (cfg.runGs) {
      if (progress && progress('gs', 0, NaN) === false) { r.canceled = true; return r; }
      timed('gs', () => gerchbergSaxton(r.phi0, inputs.illum, prop, inputs.b, inputs.mask, cfg.gs));
    }
    const op = { iters: cfg.iters, lr: cfg.lr, weights };
    timed('adam', () => optimize(r.phi0, inputs.illum, prop, inputs.b, inputs.mask, op, cb('adam')));
    if (r.canceled) return r;
    if (cfg.levels > 0) {
      const oq = Object.assign({}, op, { quant: { levels: cfg.levels, start: cfg.quantStart, ramp: cfg.quantRamp, steps: 10 } });
      timed('adam_q' + cfg.levels, () => optimize(r.phi0, inputs.illum, prop, inputs.b, inputs.mask, oq, cb('adam_q' + cfg.levels)));
    }
    (void 0, N);
    return r;
  }

  // Sweep of the field illum e^{i phi} on `grid` from 0 to z1 in nz planes, cropped to `view`
  // (sweep_volume of src/render.cpp); onPlane(p, nz) reports progress.
  function sweepField(phi, illum, iTarget, grid, z1, bandLimit, nz, view, onPlane) {
    const n = grid.n;
    const ure = new Float64Array(n * n), uim = new Float64Array(n * n);
    for (let k = 0; k < n * n; ++k) { ure[k] = illum[k] * Math.cos(phi[k]); uim[k] = illum[k] * Math.sin(phi[k]); }
    const vol = { version: 2, view, nz, pitch: grid.pitch, wavelength: grid.wavelength, z: new Float64Array(nz), amplitude: [], phase: [], charges: [],
      xz: new Float64Array(nz * view), yz: new Float64Array(nz * view), vortexCount: new Uint32Array(nz), vortexDensity: new Float64Array(nz), doePhase: null, target: null, source: null };
    for (let p = 0; p < nz; ++p) vol.z[p] = nz > 1 ? (z1 * p) / (nz - 1) : 0;
    const A = new AngularSpectrum(grid, z1, bandLimit), o = (n - view) >> 1, center = view >> 1;
    A.sweep({ re: ure, im: uim }, vol.z, (p, plane) => {
      const amp = new Float32Array(view * view), ph = new Float32Array(view * view);
      for (let i = 0; i < view; ++i)
        for (let j = 0; j < view; ++j) {
          const k = (o + i) * n + o + j, re = plane.re[k], im = plane.im[k];
          amp[i * view + j] = Math.hypot(re, im); ph[i * view + j] = wrapToPi(Math.atan2(im, re));
        }
      for (let i = 0; i < view; ++i) {
        const kx = (o + i) * n + o + center, ky = (o + center) * n + o + i;
        vol.xz[p * view + i] = plane.re[kx] ** 2 + plane.im[kx] ** 2;
        vol.yz[p * view + i] = plane.re[ky] ** 2 + plane.im[ky] ** 2;
      }
      // vortex statistics on the stored (float32) plane, as the C++ does
      const sre = new Float64Array(view * view), sim = new Float64Array(view * view);
      for (let k = 0; k < view * view; ++k) { sre[k] = amp[k] * Math.cos(ph[k]); sim[k] = amp[k] * Math.sin(ph[k]); }
      const q = vortexChargeMap({ re: sre, im: sim }, view, view);
      const q8 = new Int8Array(q.length); for (let k = 0; k < q.length; ++k) q8[k] = Math.max(-127, Math.min(127, q[k]));
      vol.amplitude.push(amp); vol.phase.push(ph); vol.charges.push(q8);
      vol.vortexCount[p] = vortexCount(q);
      vol.vortexDensity[p] = vol.vortexCount[p] / ((view - 1) * grid.pitch * (view - 1) * grid.pitch);
      if (onPlane) onPlane(p, nz);
    });
    vol.doePhase = cropCenter(phi, n, view);
    vol.target = cropCenter(iTarget, n, view);
    return vol;
  }

  // Sweep the final design: opts { nz, view, distance (0 = design distance) }.
  function sweepVolume(result, opts, onPlane) {
    const cfg = result.config, grid = result.inputs.grid;
    const nz = Math.max(opts.nz || 96, 1);
    const view = Math.min(Math.max(opts.view || Math.min(cfg.active, 384), 2), grid.n);
    const z1 = opts.distance > 0 ? opts.distance : cfg.distance;
    const final = result.runs[result.runs.length - 1];
    return sweepField(final.phi, result.inputs.illum, result.inputs.iTarget, grid, z1, cfg.bandLimit, nz, view, onPlane);
  }

  // Re-sweep a loaded volume from its v2 source block (sweep_from_source of src/render.cpp):
  // opts { nz, view, distance (0 = the stored design distance) }.
  function sweepFromSource(vol, opts, onPlane) {
    const src = vol.source;
    if (!src || !(src.n > 0)) throw new Error('sweepFromSource: the volume has no source block');
    const grid = new Grid(src.n, src.pitch, src.wavelength);
    const nz = Math.max(opts.nz || 96, 2);
    const view = Math.min(Math.max(opts.view || 8, 8), src.n);
    const z1 = opts.distance > 0 ? opts.distance : src.distance;
    const out = sweepField(Float64Array.from(src.phase), Float64Array.from(src.illum), Float64Array.from(src.target), grid, z1, src.bandLimit, nz, view, onPlane);
    out.source = src;
    return out;
  }

  function attachSource(vol, result) {
    const grid = result.inputs.grid, n = grid.n, final = result.runs[result.runs.length - 1];
    vol.source = { n, pitch: grid.pitch, wavelength: grid.wavelength, distance: result.config.distance, bandLimit: result.config.bandLimit,
      phase: Float32Array.from(final.phi), illum: Float32Array.from(result.inputs.illum), target: Float32Array.from(result.inputs.iTarget) };
    return vol;
  }

  // Settings recommended for a target (manual section 5): binary targets take the defaults
  // (TIE start, mu 0.3); continuous-tone images take the backprop start with mu 0.02. The
  // test is the fraction of mid-tone pixels (0.1 < v < 0.9 after normalizing to the maximum):
  // measured 0 for the synthetic targets and 0.19 for the rendered shell; threshold 5 %.
  function recommendConfig(target) {
    let max = 0; for (let k = 0; k < target.data.length; ++k) max = Math.max(max, target.data[k]);
    let mid = 0;
    if (max > 0) for (let k = 0; k < target.data.length; ++k) { const v = target.data[k] / max; if (v > 0.1 && v < 0.9) ++mid; }
    const midToneFraction = target.data.length ? mid / target.data.length : 0;
    const continuousTone = midToneFraction > 0.05;
    return continuousTone ? { init: 'backprop', mu: 0.02, continuousTone, midToneFraction } : { init: 'tie', mu: 0.3, continuousTone, midToneFraction };
  }

  const api = { defaults, resample, embed, cropCenter, prepare, design, sweepField, sweepVolume, sweepFromSource, attachSource, recommendConfig };
  if (isNode) module.exports = api; else { root.DOE_CORE = root.DOE_CORE || {}; Object.assign(root.DOE_CORE, api); }
  }
  const root = typeof globalThis !== 'undefined' ? globalThis : this;
  factory(root);
  if (!(typeof module !== 'undefined' && module.exports)) (root.DOE_MODULE_SOURCES = root.DOE_MODULE_SOURCES || []).push(factory.toString());
})();
